lib/common/tests/rules/pcmk__parse_type_test.c

```diff
@@ -0,0 +1,126 @@
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <stdio.h>
+
+#include <crm/common/rules_internal.h>
+#include <crm/common/unittest_internal.h>
+
+static void
+invalid(void **state)
+{
+    assert_int_equal(pcmk__parse_type("nope", pcmk__comparison_unknown,
+                                      NULL, NULL),
+                     pcmk__type_unknown);
+}
+
+static void
+valid(void **state)
+{
+    assert_int_equal(pcmk__parse_type(PCMK_VALUE_STRING,
+                                      pcmk__comparison_unknown, NULL, NULL),
+                     pcmk__type_string);
+
+    assert_int_equal(pcmk__parse_type(PCMK_VALUE_INTEGER,
+                                      pcmk__comparison_unknown, NULL, NULL),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(PCMK_VALUE_NUMBER,
+                                      pcmk__comparison_unknown, NULL, NULL),
+                     pcmk__type_number);
+
+    assert_int_equal(pcmk__parse_type(PCMK_VALUE_VERSION,
+                                      pcmk__comparison_unknown, NULL, NULL),
+                     pcmk__type_version);
+}
+
+static void
+case_insensitive(void **state)
+{
+    assert_int_equal(pcmk__parse_type("STRING", pcmk__comparison_unknown,
+                                      NULL, NULL),
+                     pcmk__type_string);
+
+    assert_int_equal(pcmk__parse_type("Integer", pcmk__comparison_unknown,
+                                      NULL, NULL),
+                     pcmk__type_integer);
+}
+
+static void
+default_number(void **state)
+{
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_lt, "1.0", "2.5"),
+                     pcmk__type_number);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_lte, "1.", "2"),
+                     pcmk__type_number);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gt, "1", ".5"),
+                     pcmk__type_number);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gte, "1.0", "2"),
+                     pcmk__type_number);
+}
+
+static void
+default_integer(void **state)
+{
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_lt, "1", "2"),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_lte, "1", "2"),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gt, "1", "2"),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gte, "1", "2"),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gte, NULL, NULL),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gte, "1", NULL),
+                     pcmk__type_integer);
+
+    assert_int_equal(pcmk__parse_type(NULL, pcmk__comparison_gte, NULL, "2.5"),
```

Comment:
> > Still, it seems better for `pcmk__parse_type()` to check whether `'.'` is in each value individually, so that we don't return false prematurely in this case.
> 
> Good point
> 
> > As an aside, this is the ONLY caller of `pcmk__char_in_any_str()`. So if we remove it from `pcmk__parse_type()`, then we can drop it.
> 
> It's separate mainly for unit testing

But of course we won't need it at all anymore